src/coreclr/jit/lower.cpp

```diff
@@ -5753,6 +5753,50 @@ void Lowering::LowerShift(GenTreeOp* shift)
         shift->gtOp2->ClearContained();
     }
     ContainCheckShiftRotate(shift);
+
+#ifdef TARGET_ARM64
+    // Check if we can recognize ubfiz/sbfiz idiom in LSH(CAST(X), CNS) pattern
+    if (comp->opts.OptimizationEnabled() && shift->OperIs(GT_LSH) && shift->gtGetOp1()->OperIs(GT_CAST) &&
+        shift->gtGetOp2()->IsCnsIntOrI() && !shift->isContained() && varTypeIsIntegral(shift))
+    {
+        GenTreeIntCon* cns  = shift->gtGetOp2()->AsIntCon();
+        GenTreeCast*   cast = shift->gtGetOp1()->AsCast();
+
+        if (!cast->isContained() && !cast->IsRegOptional() && !cast->gtOverflow() &&
+            // CastOp should not be a small int, e.g. CAST(IND(ByteTarget))
+            cast->CastOp()->TypeIs(TYP_LONG, TYP_INT))
+        {
+            // Cast can be either "TYP_LONG <- TYP_INT" or "TYP_INT <- %SMALL_INT% <- TYP_INT"
+            UINT32 treeSize = genTypeSize(cast);
+            UINT32 fromSize = genTypeSize(cast->CastFromType());
+            UINT32 toSize   = genTypeSize(cast->CastToType());
+
+            assert(!cast->CastOp()->isContained());
+
+            if (treeSize >= fromSize) // It has to be an upcast
+            {
+                if (toSize < fromSize)
+                {
+                    // Special case: "TYP_INT <- %SMALL_(U)INT% <- TYP_INT"
+                    assert(varTypeIsSmall(cast->CastToType()) && (fromSize == 4));
+                    toSize = 4;
+                }
+                assert(toSize >= 4);
+
+                // LSH(TYP_LONG <- TYP_INT, CNS) => 1..31
+                // LSH(TYP_INT <- TYP_(U)SHORT <- TYP_INT, CNS) => 1..15
+                // LSH(TYP_INT <- TYP_(U)BYTE <- TYP_INT, CNS) => 1..7
+                if ((UINT32)cns->IconValue() < (toSize * 4))
+                {
+                    // cns is small and should already be contained at this point
+                    assert(cns->isContained());
+                    JITDUMP("Recognized ubfix/sbfix pattern in LSH(CAST, CNS), marking CAST node as contained.");
+                    MakeSrcContained(shift, cast);
+                }
```

Comment:
This logic is a bit unclear... I think we're guarding against overshift, here, and so `toSize * 4` is more like `(toSize * BITS_PER_BYTE) / 2` right? The comment appears to suggest we can have this value tuned specifically for small types (because they allow more "headroom" in shifting, so to speak), but then `toSize` is always `4` or `8`?

It also appears that `toSize` is always equal to `treeSize`?

I don't know, maybe it would be clearer to reformulate this in terms of how many bits are we going to extract and how many are we going to shift, something like:
```c++
size_t srcBytes  = varTypeIsSmall(cast->CastToType()) ? genTypeSize(cast->CastToType())
                                                      : genTypeSize(cast->CastOp());
size_t srcBits   = srcBytes * BITS_PER_BYTE;
size_t shiftBits = cns->IconValue();
size_t dstBits   = genTypeSize(cast) * BITS_PER_BYTE;

if ((shiftBits <= 64) && (srcBits + shiftBits) <= dstBits) { /* Yay! */ }
```

Reply:
good idea, thanks!